Adding more retriever examples to the documentation examples (#116196)
elastic/elasticsearch: docs/reference/search/search-your-data/retrievers_examples.asciidoc

that would operate on top of 2 other retrievers, a `knn` and a `standard`.  Our query in this case would look something like the following:

[source,js]
----
GET /retrievers_example/_search
{
   "retriever":{
       "rrf": {
           "retrievers":[
               {
                   "standard":{
                       "query":{
                           "semantic":{
                               "field": "inference_field",
                               "query": "Can I use generative AI to identify user intent and improve search relevance?"
                           }
                       }
                   }
               },
               {
                   "knn": {
                       "field": "vector",
                       "query_vector": [0.23, 0.67, 0.89],
                       "k": 3,
                       "num_candidates": 5
                   }
               }
           ],
           "rank_window_size": 10,
           "rank_constant": 1
       }
   },
   "_source": ["text", "topic"]
}
----
//NOTCONSOLE
end::basic-rrf-retriever-with-semantic-query[]

[discrete]

> tag::rrf-retriever-with-collapse[]

Reviewer: discrete tag is in wrong spot

=== Example: Grouping results by year with `collapse`

> Let's say that we have our results, but we get back many documents for the same `year`. We can now use `collapse` with retrievers, to group results based

Reviewer: ```suggestion

In our result set, we have many documents with the same `year` value. We can clean this
up using the `collapse` parameter with our retriever. This enables grouping results by
any field and returns only the highest-scoring document from each group. In this example
we'll group by year.
```

[source,js]
----
GET /retrievers_example/_search
{
   "retriever":{
       "rrf": {
           "retrievers":[
               {
                   "standard":{
                       "query":{
                           "semantic":{
                               "field": "inference_field",
                               "query": "Can I use generative AI to identify user intent and improve search relevance?"
                           }
                       }
                   }
               },
               {
                   "knn": {
                       "field": "vector",
                       "query_vector": [0.23, 0.67, 0.89],
                       "k": 3,
                       "num_candidates": 5
                   }
               }
           ],
           "rank_window_size": 10,
           "rank_constant": 1
       }
   },
   "collapse": {
       "field": "year",
       "inner_hits": {
           "name": "topic related documents",
           "_source": ["text", "year"]
       }
   },
   "_source": ["text", "topic"]
}
----
//NOTCONSOLE
end::rrf-retriever-with-collapse[]

tag::rrf-on-top-of-semantic-reranker[]
[discrete]
=== Example: RRF with semantic reranker

> For this scenario, let's say that we want to swap our semantic query with our `my-awesome-rerank-model` reranker that we
> have already setup. The main difference now is that, since this is a reranker, it will need an initial pool of docs to rerank!
> We know that we want to work with `ai` topics, so let's try to do just that!

Reviewer: ```suggestion

For this example, we'll replace our semantic query with the `my-awesome-rerank-model`
reranker we previously configured. Since this is a reranker, it needs an initial pool of
documents to work with. In this case, we'll filter for documents about `ai` topics.
```

Reviewer: I love the enthusiasm, but exclamation marks can become visually distracting :)

                        "field": "vector",
                        "query_vector": [
                            0.23,
                            0.67,
                            0.89
                        ],
                        "k": 3,
                        "num_candidates": 5
                    }
                },
                {
                    "text_similarity_reranker": {
                        "retriever": {
                            "standard": {
                                "query": {
                                    "term": {
                                        "topic": "ai"
                                    }
                                }
                            }
                        },
                        "field": "text",
                        "inference_id": "my-awesome-rerank-model",
                        "inference_text": "Can I use generative AI to identify user intent and improve search relevance?"
                    }
                }
            ],
            "rank_window_size": 10,
            "rank_constant": 1
        }
    },
    "_source": [
        "text",
        "topic"
    ]
}

> 

Reviewer: ```suggestion
```

----
//NOTCONSOLE
end::rrf-on-top-of-semantic-reranker[]

tag::text-similarity-reranker-on-top-of-rrf[]
[discrete]
=== Example: Combine semantic reranker with RRF

> In the example above, we had a `text_similarity_reranker` retriever within an `rrf` one, but remember that retrievers support full
> composability, so we can rerank the results of an rrf retriever. Let's try to do this with the query from the first example above

Reviewer: ```suggestion
Previously, we used a `text_similarity_reranker` retriever within an `rrf` retriever.
However, retrievers support full composability, so we can also rerank the results of an
`rrf` retriever. Let's apply this to our first query example.

```

                                   }
                               }
                           }
                       },
                       {
                           "knn": {
                               "field": "vector",
                               "query_vector": [
                                   0.23,
                                   0.67,
                                   0.89
                               ],
                               "k": 3,
                               "num_candidates": 5
                           }
                       }
                   ],
                   "rank_window_size": 10,
                   "rank_constant": 1
               }
           },
           "field": "text",
           "inference_id": "my-awesome-rerank-model",
           "inference_text": "What are the state of the art applications of AI in information retrieval?"
       }
   },
   "_source": ["text", "topic"]
}

----
//NOTCONSOLE
end::text-similarity-reranker-on-top-of-rrf[]

tag::chaining-text-similarity-reranker-retrievers[]
[discrete]
=== Example: Chaining multiple semantic rerankers

> Full composability, means that we can also chain together multiple retrievers of the same type. Say that we have another
> very computationally expensive reranker that is more fine-grained for AI content. We can now also rerank the results of a `text_similarity_reranker`,
> using another `text_similarity_reranker` retriever, which could operate on different fields and/or inference services!

Reviewer: ```suggestion
Full composability means we can chain together multiple retrievers of the same type. For instance, imagine we have a computationally expensive reranker that's specialized for AI content. We can rerank the results of a `text_similarity_reranker` using another `text_similarity_reranker` retriever. Each reranker can operate on different fields and/or use different inference services.
```

{
   "retriever": {
       "text_similarity_reranker": {
           "retriever": {
               "text_similarity_reranker": {
                   "retriever": {
                       "knn": {
                           "field": "vector",
                           "query_vector": [
                               0.23,
                               0.67,
                               0.89
                           ],
                           "k": 3,
                           "num_candidates": 5
                       }
                   },
                   "rank_window_size": 100,
                   "field": "text",
                   "inference_id": "my-awesome-rerank-model",
                   "inference_text": "What are the state of the art applications of AI in information retrieval?"
               }
           },
           "rank_window_size": 10,
           "field": "text",
           "inference_id": "my-other-more-expensive-rerank-model",
           "inference_text": "Applications of Large Language Models in technology and their impact on user satisfaction"
       }
   },
   "_source": [
       "text",
       "topic"
   ]
}
----
//NOTCONSOLE

> Note that in the example above, we initially rerank the top 100 documents from the `knn` search using the `my-awesome-rerank-model` reranker,
> and then pick the top 10 results and rerank them using the more fine-grained `my-other-more-expensive-rerank-model`.

Reviewer: ```suggestion
Note that our example applies two reranking steps. First, we rerank the top 100
documents from the `knn` search using the `my-awesome-rerank-model` reranker. Then we
pick the top 10 results and rerank them using the more fine-grained
`my-other-more-expensive-rerank-model`.
```

end::chaining-text-similarity-reranker-retrievers[]

tag::rrf-retriever-with-aggs[]
[discrete]
=== Example: Combine RRF with aggregations

> We have seen some examples with retrievers' composability, but we can also now support most of the standard search functionality!
> Let's say that we want to compute aggregations for the `rrf` retriever. Note that the aggregations
> in a compound retriever will be computed based on the nested retrievers it holds. So this means that for the following query

Reviewer: ```suggestion
Retrievers support both composability and most of the standard `_search` functionality. For instance,
we can compute aggregations with the `rrf` retriever. When using a compound retriever,
the aggregations are computed based on its nested retrievers. Here's an example:
```

Reviewer: this text seemed to end abruptly, consider adding any missing context!

Author: yeap, missed this when I was moving content around. added in c11c47f8e7373e316f029d52a4dbf4df5a7905fa

tag::basic-rrf-retriever-with-semantic-query[]
[discrete]

> === Combining kNN and semantic search with RRF

Reviewer 2: Shouldn't it be more interesting to combine BM25 with semantic / kNN for hybrid search?

Reviewer 2: Good point - IMO having something in multiple places is OK as long as it is relevant on the current topic and can be accessed through different use cases (RRF is specific, retrievers is more of a framework). 

For me it doesn't make a lot of sense to combine semantic query with knn - they are basically different ways of achieving vector search, I'm not sure if I would use it in the Real World ™️ . But hybrid is something I definitely would.

Author: I see what you mean. Updated to have a `standard` retriever with a `query_string` query and a `knn` retriever.

Author: Will also remove the initial ELSER setup now that we don't directly reference the `semantic` query anymore.

                       "query":{
                           "semantic":{
                               "field": "inference_field",
                               "query": "Can I use generative AI to identify user intent and improve search relevance?"
                           }
                       }
                   }
               },
               {
                   "knn": {
                       "field": "vector",
                       "query_vector": [0.23, 0.67, 0.89],
                       "k": 3,
                       "num_candidates": 5
                   }
               }
           ],
           "rank_window_size": 10,
           "rank_constant": 1
       }
   },
   "_source": ["text", "topic"]
}
----
//NOTCONSOLE
end::basic-rrf-retriever-with-semantic-query[]

tag::rrf-retriever-with-collapse[]
[discrete]
=== Grouping results by year with `collapse`

In our result set, we have many documents with the same `year` value. We can clean this
up using the `collapse` parameter with our retriever. This enables grouping results by
any field and returns only the highest-scoring document from each group. In this example
we'll group by year.

> on a given value and pick just the top for each sub-group!

Reviewer 2: Is this a leftover?

```suggestion
```

Author: Yeap, due to incorrect merging. Reworked the sentence to 

```
In this example we'll collapse our results based on the `year` field.
```

[source,js]
----
GET /retrievers_example/_search
{
   "retriever":{
       "rrf": {
           "retrievers":[
               {
                   "standard":{
                       "query":{
                           "semantic":{
                               "field": "inference_field",
                               "query": "Can I use generative AI to identify user intent and improve search relevance?"
                           }
                       }
                   }
               },
               {
                   "knn": {
                       "field": "vector",
                       "query_vector": [0.23, 0.67, 0.89],
                       "k": 3,
                       "num_candidates": 5
                   }
               }
           ],
           "rank_window_size": 10,
           "rank_constant": 1
       }
   },
   "collapse": {
       "field": "year",
       "inner_hits": {
           "name": "topic related documents",
           "_source": ["text", "year"]
       }
   },
   "_source": ["text", "topic"]
}
----
//NOTCONSOLE
end::rrf-retriever-with-collapse[]

tag::rrf-on-top-of-semantic-reranker[]
[discrete]
=== RRF with semantic reranker

> For this example, we'll replace our semantic query with the `my-awesome-rerank-model`

Reviewer 2: I see you have semantic reranker on top of the previous two retrievers in the following example. For me it would make more sense to have it as the first example:

- It makes more sense IMO as we're reranking as part of a final stage on top
- We build directly on top of the previous example
- We build real-world retrievers usage first by addressing hybrid search and reranking first

Then you can keep exploring the full composability in the following examples.

Author: So would you suggest to remove the `RRF with semantic reranker` example, or change the order to something like the following? 
* `Rerank results of RRF retriever` (building on previous example)
* `RRF with semantic reranker`.  (swapping the rrf with the semantic reranker retrievers)
* `Chaining multiple semantic rerankers`

Author: Updated order to: 
```
include::retrievers_examples.asciidoc[tag=basic-rrf-retriever-with-semantic-query]
include::retrievers_examples.asciidoc[tag=rrf-retriever-with-collapse]
include::retrievers_examples.asciidoc[tag=text-similarity-reranker-on-top-of-rrf]
include::retrievers_examples.asciidoc[tag=rrf-on-top-of-semantic-reranker]
include::retrievers_examples.asciidoc[tag=chaining-text-similarity-reranker-retrievers]
include::retrievers_examples.asciidoc[tag=rrf-retriever-with-aggs]
```
